macOS opengl-cb backend

Akemi opened the pull request:
this is a WIP and attempts to fix various annoying macOS bugs and the very long standing fullscreen regression completely. it still needs testing and some fix-ups.

this will definitely fix: #5478, #5393, #5152, #5151, #4615, #4476, #3746, #3739, #2392, #2217
and possibly fixes: #3978

feel free to test and report unknown bugs. for now one just needs to set `--vo=opengl-cb`, ~~though if the mpv core is too fast you might get a "No context set." error.~~

there are a few things to do where i don't know how to solve them yet and i might need some help.
TODO cocoa-cb:
- ~~proper cocoa-cb init routine, (--no-config --vo=opengl-cb fails since no context yet)~~
- ~~cocoa-cb vo option or comparable~~ (just compile time flag, uses vo=opengl-cb)
- ~~doesn't draw when minimised, off screen, different space etc.~~
- ~~get a hide (and unhide) event for windowless playback~~
- ~~remove "[opengl-cb] icc-profile-auto is not available with opengl-cb" warning in an acceptable way~~
- ~~change lux scale~~ (db0fb3c48b12f09170231fccd2c91687d9ce21db)
- ~~remove "Taking screenshot failed." warning, appears on screenshot (window), not too happy with the window screenshot code~~
- ~~on shutdown wait for animations to finish~~
- ~~[performance data buggy](https://github.com/mpv-player/mpv/issues/4615#issuecomment-326738862)~~
- ~~ICC profile switching leads to [flicker](https://0x0.st/74V.mov) in some cases (with icc-profile-auto=no, most likely an Apple bug)~~

TODO Build:
- ~~flag to deactivate cocoa-cb~~
- ~~dynamic developer path for static_swift~~
- ~~set build target min system requirements~~

things i will fix soonish:
- ~~fullscreen animation stretches video to target aspect ratio~~
- ~~crashes on video stream switch to no-video (related to "hide and unhide event for windowless")~~
- ~~implement none-native fs~~
- ~~add license headers~~
- ~~(deactivate the early flush on macOS)~~

TODO misc:
- ~~add a deprecation warning for the old opengl cocoa backend?~~

wm4 commented on video/out/vo_opengl_cb.c:
```diff
@@ -470,9 +507,20 @@ static int control(struct vo *vo, uint32_t request, void *data)
+    int r = VO_NOTIMPL;
+    if (p->ctx->control_cb) {
+        int events = 0;
+        r = p->ctx->control_cb(p->ctx->control_cb_ctx, &events, request, data);
```
Unsynchronized access to the callback.

wm4 commented on player/screenshot.c:
```diff
@@ -454,6 +454,18 @@ void screenshot_to_file(struct MPContext *mpctx, const char *filename, int mode,
+        write_screenshot(mpctx, img, filename, NULL, true);
+    talloc_free(filename);
+    talloc_free(img);
+}
```
I don't really like this approach, and also synchronization of any form is missing.

AirPort commented:
As I said in the previous issue, I tested it for a bit and so far I didn't notice any particular difference performance wise compared to master, and it definitely seems to fix the mentioned issues (not 100% sure about #3978, only time and use will tell since it's difficult to reproduce at will).
I'll keep on testing and report back if I notice something worthy of note.

Just one question though: during testing I once forgot to type --vo=opengl-cb, so it started with regular vo=opengl and hanged without drawing any window. Are the two cocoa backends mutually exclusive or in the final version will be still possible to select and use the current one (e.g. for testing purposes)?

Akemi commented:
that will be fixed in the final version. it's just a matter of a proper option and updating the init routine of it (a little bit). after that using both will be possible.

JCount commented:
This may not be useful, but on OS X 10.9 and later, the `CGLOpenGLProfile` `kCGLOGLPVersion_GL4_Core` could potentially be useful in instances where the `cocoa-cb` would want (the system to) `/* choose a renderer capable of GL4.1 or later   */`.

Akemi commented:
kCGLOGLPVersion_3_2_Core is just a misnomer atm and just means request the highest possible version but at least 3.2. so yeah kCGLOGLPVersion_GL4_Core isn't useful at all atm.

JCount commented:
Ok, I wasn't sure if some of the dual GPU systems could have only one GPU capable of OpenGL 4.1, which if not already selected as the default, this could get the system to switch over to that GPU. Essentially, would it ever be desirable to raise the minimum bound that high for the `CGLOpenGLProfile`. 😄

ottob commented:
Thanks for working on this. I tried to build d5c93df on high sierra (17A360a) but it fails with:
```
Waf: Leaving directory `/Users/ottob/Documents/Develop/mpv-build/mpv/build'
Build failed
 -> task in 'osdep/macOS_swift.o osdep/macOS_swift.h osdep/macOS_swift.swiftmodule' failed with exit status 1 (run with -v to display more information)
```
Any ideas?

Akemi commented:
if you can run the build with `-v` and post that log it would be helpful.

AirPort commented:
It's the same error I had when I tried to build it without the full Xcode.
You can try a temprary workaround by manually changing the 
`'-L/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift_static/macosx'`
like in wscript_build.py to 
`'-L/Library/Developer/CommandLineTools/usr/lib/swift_static/macosx'`
and see if it builds.

ottob commented:
Thanks for the hint AirPort. I modified wscript_build.py file like so (I have xcode 9 beta installed)
```diff
-            '-L/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift_static/macosx',
+            '-L/Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift_static/macosx',
```
But still it wont build, looks like it might be a Swift 4 error. I have put the verbose log here: https://gist.github.com/ottob/11648da514c0986f72b2c7a60fceb9cf

macdavis commented:
It seems it just produces a black screen when mpv's dithering is enabled. Tested with d5c93df. 
Log file: [http://sprunge.us/YARh](http://sprunge.us/YARh)

AirPort commented:
I can reproduce the black screen, but only with `--dither-size-fruit=8`. Any other value (included default 6) seems to work fine.

Akemi commented:
@ottob yeah i tested this with swift 3. you can add a `--swift-version 3` [here](https://github.com/Akemi/mpv/blob/macOS_cocoa-cb2/wscript_build.py#L132), behind the -O. that should work, couldn't test it myself.

@macdavis dithering is always enabled by default. it's just the size of the fruit dither you used. it already takes ages to load here with the normal cocoa backend. it also does work with cocoa-cb you just have to wait (the same time) till the video appears. the difference is that with the normal cocoa backend the mpv core blocks till it's done computing, with opengl-cb it can already start drawing without the blocking, that results in a temporary black screen.

macdavis commented:
Thanks! The video does show after some time of waiting.

Akemi commented:
@macdavis i reported your problem as a separate issue #4852, since it's a problem all opengl-cb users face.

ottob commented:
> yeah i tested this with swift 3. you can add a --swift-version 3 here, behind the -O. that should work, couldn't test it myself.

`-swift-version 3` single dash worked, thanks.

Akemi commented:
indeed a single dash. i blame mpv for that.

ottob commented:
Ive only done some very basic testing, but have not noticed any regressions. Thanks for maintaining the macOS code :)

Some debug prints are still around, but you probably know that:
```
cocoa_set_mpv_handle start
setMpvHandle start
copyCGLPixelFormat start-----
copyCGLPixelFormat end-----
copyCGLContext------
setMpvHandle end
cocoa_set_mpv_handle end
```

wm4 commented on video/out/vo_opengl_cb.c:
```diff
@@ -479,7 +484,13 @@ static int reconfig(struct vo *vo, struct mp_image_params *params)
 
 void mp_client_set_icc_profile(struct mpv_opengl_cb_context *ctx, bstr icc_data)
 {
+    pthread_mutex_lock(&ctx->icc_lock);
```
Would it be possible to always call the function on the renderer thread? That's probably also how an official opengl-cb API for this would work.

Akemi replied:
oh do you mean like it was [before](https://github.com/mpv-player/mpv/pull/4833/commits/ad89cb84b59216fbf1906b41972b67e211ce34fc#diff-b5010a29091b72c707c97eae67a97254L80) (thought the current approach was preferable) or do it in the mpv_opengl_cb_draw function via a flag?

wm4 replied:
No, just rule that `mp_client_set_icc_profile` can be called in the GL thread only, so that no additional synchronization is needed.

Akemi replied:
sure can do that. something broke anyway when i rebased it with the current "vo_gpu" changes. just getting a "blue screen" and shader errors on some icc profile changes.

should i mention that rule somewhere?

wm4 replied:
As a comment on the function would be good.

Akemi commented:
hopefully i fixed the last 2 segfaults that were my fault. there is still one left that is most likely related to #4988 (http://sprunge.us/FHXV) or is exactly the same. i will clean up the code soonish and hopefully will be done at the end of the week.

Akemi commented:
as far as the code goes i am finally done. i will squash everything laters. i am not overly confident in my build changes maybe, they probably need changes.

AirPort commented:
I just gave it a try, but it gives my building errors: https://0x0.st/szoL.txt

Akemi commented:
thx, should be fixed now.

AirPort commented:
Yep, it builds now. As far as I can see, I can confirm my early observations: I don't see any glaring performance issue, those long standing little issues seems to be gone and I failed to see any new one in my (admittedly quick) tests. 
Great work!

(Just out of curiosity, why did you remove the window screenshot code? Too fragile?)

Akemi commented:
sadly it doesn't work how mpv expects the window screenshot to work and is always async (forces it even). for example scripts that rely on screenshots being sync won't work anymore. it was a 'special' mechanism just for that purpose. 
i still plan to add it properly so it works with the mpv mechanism. i have a working change [here](https://github.com/Akemi/mpv/commit/8634b46cd43d1ff99251897260c53c0e18b673a7), though it is possible that it deadlocks completely in some cases. i still need to test it.

torque commented:
Cycling video (default keybinding is `_` ) to none with these changes causes a segfault in `vo_calc_window_geometry2` because the argument `vo` is passed as null to the function.

```
frame #0: 0x00000001000d2d9d mpv`vo_calc_window_geometry2(vo=0x0000000000000000, screen=0x00007ffeefbfdf90, dpi_scale=2, out_geo=0x00007ffeefbfdfa0) at win_state.c:82 [opt]
   79  	void vo_calc_window_geometry2(struct vo *vo, const struct mp_rect *screen,
   80  	                              double dpi_scale, struct vo_win_geometry *out_geo)
   81  	{
-> 82  	    struct mp_vo_opts *opts = vo->opts;
   83
   84  	    *out_geo = (struct vo_win_geometry){0};
   85
```

Akemi commented:
thx, fixed it.

macdavis commented:
mpv bundle cannot be opened with `-vo=opengl-cb`. The crash log: [http://sprunge.us/CaZA](http://sprunge.us/CaZA)

I also noticed that with the recent changes in mpv, opengl-cb will create the hardware decoding OpenGL interop context by default and there is no way to disable it. (Tried `-gpu-hwdec-interop=no`, but not working)

Akemi commented:
> mpv bundle cannot be opened with -vo=opengl-cb. The crash log: http://sprunge.us/CaZA

works here as intended. 

> I also noticed that with the recent changes in mpv, opengl-cb will create the hardware decoding OpenGL interop context by default and there is no way to disable it. (Tried -gpu-hwdec-interop=no, but not working)

no idea how this is related to this PR? it's a opengl-cb [limitation](https://github.com/mpv-player/mpv/blob/643a1fc7de84b234d5a4723fa0ec31f9351e770b/libmpv/opengl_cb.h#L110). does it cause any problems?

Akemi commented:
this is just me poking in the dark. does this commit fix your first issue with the bundle @macdavis?

macdavis commented:
Thanks! Commit 199ca0b indeed fixes my problem. I also find another issue. When you first open the mpv bundle and drag a music file that attaches an artwork into mpv, the image aspect ratio is wrong. Sample file: [https://0x0.st/sH1l.m4a](https://0x0.st/sH1l.m4a)

<img src="http://ww1.sinaimg.cn/large/8d5749ccgy1fm7elq6iffj213q151q4q.jpg"/>

It should look like
<img src="http://ww1.sinaimg.cn/large/8d5749ccgy1fm7en5d1q9j208c08cmwz.jpg"/>

As for OpenGL interop context, I don't see any particular problem. I was just confused and didn't know that was expected in opengl-cb. Thanks for your clarification. I tested in the old opengl cocoa backend as well and found OpenGL interop cannot be created even if `gpu-hwdec-interop=auto` is set. But I think it's not relevant to this PR and the old cocoa backend is going to be deprecated anyway.

Akemi commented:
this should fix it.

macdavis commented:
Thanks!  327b83e did fix the previous issue. There is another issue with the non-native fullscreen. Firstly, the title bar is always visible. Secondly, window/fullscreen switch for several times (5-10 times) can also cause wrong aspect ratio. Such as,
<img src="http://ww1.sinaimg.cn/large/8d5749ccgy1fm8y7dyqorj22801e0gw9.jpg"/>
and
<img src="http://ww1.sinaimg.cn/large/8d5749ccgy1fm8y847njxj213u150tdn.jpg"/>

In addition, `ontop-level=system` seems not working in OpenGL-cb

Akemi commented:
fixed the wrong aspect ratio problem. i can't reproduce the 'titlebar always visible problem' or the 'ontop-level=system' issue, they both work as expected for me. could you pls provide additional info like mentioned in the issue template. also the minimal steps with `--no-config` would probably be helpful.

macdavis commented:
1. `mpv -no-config -no-native-fs -fs -vo=opengl-cb`. The title bar is not automatically hidden, as shown in the following fullscreen screenshot. [http://ww1.sinaimg.cn/large/8d5749ccgy1fm9l47uepuj22801e0gue.jpg](http://ww1.sinaimg.cn/large/8d5749ccgy1fm9l47uepuj22801e0gue.jpg)

2. `mpv -no-config -vo=opengl-cb -ontop -ontop-level=system`. The window is still behind the dock. http://ww1.sinaimg.cn/large/8d5749ccgy1fm9l6rnu7pj20jo0fe794.jpg. 
While without opengl-cb (`mpv -no-config -ontop -ontop-level=system`) The window is on top of the dock, which is as expected. http://ww1.sinaimg.cn/large/8d5749ccgy1fm9l95hgp3j20j60jctbe.jpg


3. There is another crash, which happened pretty random. [http://sprunge.us/NZAH](http://sprunge.us/NZAH)

Akemi commented:
1. fixed, forgot to port 2 lines from the old code.
2. fixed, this was a regression from my init method rewrite. it actually set the ontop level but only for the preallocated window and not the new one.
3. i don't know yet what the problem for this is, only have a guess. did this happen on init/open mpv? any additional info you can give me?

macdavis commented:
It did happen on init/open mpv by clicking the icon in the dock. But I can't conclude an exact way to reproduce the crash. 

There is also a little defect in `ontop-level=system`, which affects both old OpenGL backend and opengl-cb. With `mpv -no-config -ontop-level=system -ontop -fs -vo=opengl-cb`, if you move the cursor to the very top edge of the screen, you'll see the menu bar (in my case, the Wi-Fi indicator) and title bar.
http://ww1.sinaimg.cn/large/8d5749ccgy1fm9os90btlj22801e0n66.jpg

macdavis commented:
6d18c94 fixed the menu bar, but still showing the title bar.
http://ww1.sinaimg.cn/large/8d5749ccgy1fm9po1l1iyj22801e07da.jpg

Akemi commented:
i see. then i probably know what happens. it might be a bit tricky to fix and confirm it is actually fixed.

seems like the icons from the menubar are actually on the `statusWindow` window level. i just moved system window level one up so it also hides the icons in native-fs. that the titlebar is showing is by design (Apple) we can't hide it. it's always above our window, even above `maximumWindow` level. there might be a hack for this, but then we should ask ourselves should we even use `ontop` in fullscreen? it was/is only useful for one case in the none-native fs i believe. we also can't move the titlebar to the where the menubar is hidden, since it is also like this by design (Apple).

Akemi commented:
there is a potential fix for the mouseMoved segfault. though i couldn't reproduce the problem even once.

macdavis commented:
I'm not sure whether it is technically feasible, but could you please further perfect the fullscreen animation mechanism? 

1. Entering fullscreen
Currently, the title bar will not disappear until switching to fullscreen space is ALMOST done. The moment the title bar is gone, the video will continue expanding to fullscreen without any scaling animation, which looks unnatural. 
As for other application, such as Finder and Terminal, the title bar will disappear as soon as fullscreen is triggered. Without the title bar, the fullscreen animation looks much smoother.

2. Exiting fullscreen
There is no scaling animation at all by clicking the green fullscreen bottom on the title bar. If you exit fullscreen by keyboard shortcut "F", the title bar will suddenly appear after the scaling animation. While in Terminal, the title bar comes with the smooth scaling animation. For short, it seems the order of title bar's appearing/disappearing and scaling animation is reversed during entering/exiting fullscreen.

Akemi commented:
QuickTime for example behaves like mpv. titlebar is showing till the end of the animation and only appears again at the end of the animation when exiting the fullscreen. the only difference is that the titlebar is inside the window and not outside. i am planning to change that in the future to prevent some other macOS bugs too, so i don't think me trying to fixing that now is feasible. as for now the hiding and showing of the titlebar are done by the cocoa provided fullscreen API function, so it's not easily changeable. in your examples (finder and terminal) they either have a custom titlebar or their animation is done with a 'buffered screenshot' of the actual windows and a cross-fade, whereas mpv actually animates the content of the window (which also isn't an easy feat tbh). 

> There is no scaling animation at all by clicking the green fullscreen bottom on the title bar.

that shouldn't be and works here as intended. the window is always animated and clicking the green circle or hitting the f button should make no difference.

rodlie commented:
Hi, 
there are build issues on 10.11.6 (Xcode 8.2.1).

`<unknown>:0: error: unknown argument: '-swift-version'`

Removing the argument results in

`In file included from ../osdep/macosx_events.m:43:
./osdep/macOS_swift.h:238:40: error: cannot find protocol declaration for 'CALayerDelegate'; did you mean 'NSDrawerDelegate'?
@interface VideoLayer : CAOpenGLLayer <CALayerDelegate>
                                       ^~~~~~~~~~~~~~~
                                       NSDrawerDelegate
/System/Library/Frameworks/AppKit.framework/Headers/NSDrawer.h:91:11: note: 'NSDrawerDelegate' declared here
@protocol NSDrawerDelegate <NSObject>
          ^
1 error generated.`

Akemi commented:
i am inclined to not support 10.11 in that case. it's quite a pain in the ass that 10.11 only gets swift 3.0 and therefor doesn't have the `-swift-version` flag from 3.1. 

the latter could be worked around. though i wonder the [delegate](https://developer.apple.com/documentation/quartzcore/calayer/1410984-delegate?language=objc) property apparently is available since 10.5 but the [CALayerDelegate](https://developer.apple.com/documentation/quartzcore/calayerdelegate?language=objc) protocol was only added with 10.12.

Argon- commented:
In that case, as a 10.11 user that probably won't go to 10.13 for some time*, I hope the current backend's deprecation phase will last long enough. :p

* ultimately I do intend to upgrade but not at the current state of this OS

Akemi commented:
the first problem is just a compile time problem because 10.11 doesn't support the most recent dev environment from Apple (but yeah you can just patch that one argument out if you compile it yourself anyway). alternatively one can 'just' compile on a newer system for 10.11. maybe we could check those sorts of things on configure? maybe some thing need to be moved to configure anyway?

the latter problem can be circumvented by a few lines of additional code (~3) i believe. so we could make the code compatible with 10.11. actually swift would make this a lot more convenient with macOS versions checks.

i only really planned to deprecated the old backend but not remove it (any time soon). i just don't plan on fixing bugs for it anymore. though it could still be useful for comparing it to cocoa-cb in the future.

on a side note, why do you stay on 10.11 one of the worst versions since 10.10 imo?

Argon- commented:
It's actually the most stable for me. Also with the current cocoa backend, all I need is opengl early flush turned off and I have zero of the bugs people often report on OS X with the exception of a flickering volume indicator (and I'm not bothered by that). I have multiple computers (a hackintosh among them) that are continuously synchronized replicas (using unison), so I prefer to update them in lockstep with each other and currently I don't consider it worth the effort as really the only worthwhile feature are more recent dev tools. 
I'm well aware I cannot stay on 10.11 and do intend to upgrade (especially because of dev tools) but I think I better wait until a few more 10.13 updates hit the public before I let this bug infested security nightmare hit my hardware. :s

macdavis commented:
During the fullscreen animation, some particular frame looks wired. This happens when I continuously cycle fullscreen/window.

Most of the time, the frame looks normal, as shown in [this video](https://0x0.st/sX_v.mov). 

Bug 1: During the exit animation, one frame becomes 1/4 size of that should have been, as shown in [this video](https://0x0.st/sX_w.mov). 

Bug 2: During the fullscreen animation, the size one frame becomes 4 times larger then it should have been, as shown in [this video](https://0x0.st/sX_x.mov).

Akemi commented:
@rodlie i pushed a potential build fix for 10.11, though you still need to remove the `-swift-version` manually for now. i tried it in a VM but it got a bunch of linker errors.

Akemi commented:
@macdavis i kinda know about this problem. it's related to the black flickering. this is just me assuming now since i couldn't find the root of the problem yet. i believe there is some event that rebuilds the window hierarchy and that resets some things. if mpv redraws at that exact moment we get that glitch/glitches. the problem i have is that i couldn't isolate that event and subsequently can't tell mpv to stop drawing.

macdavis commented:
Another crash during opening mpv bundle. Report: [https://0x0.st/sXnD.txt](https://0x0.st/sXnD.txt)

wm4 commented on video/out/opengl/context.c:
```diff
@@ -295,6 +297,7 @@ bool ra_gl_ctx_submit_frame(struct ra_swapchain *sw, const struct vo_frame *fram
     case FLUSH_YES:
         gl->Flush();
     }
+#endif
```
I'd just set `gl->Flush` to a dummy function instead. Does it still cause issues with opengl-cb?

Akemi replied:
i didn't have any issues anymore with it but @AirPort still had some issues (https://github.com/mpv-player/mpv/issues/4615#issuecomment-326767362) with his integrated GPU.

wasn't too sure how to handle the overwriting of gl->Flush() since i have to use it [here](https://github.com/mpv-player/mpv/blob/master/video/out/opengl/context_cocoa.c#L161). including the header file for the function leads to some compile warning:
```
../video/out/opengl/gl_headers.h:614:9: warning: 'GL_DRAW_FRAMEBUFFER_BINDING' macro redefined [-Wmacro-redefined]
#define GL_DRAW_FRAMEBUFFER_BINDING       0x8CA6
        ^
/System/Library/Frameworks/OpenGL.framework/Headers/glext.h:360:9: note: previous definition is here
#define GL_DRAW_FRAMEBUFFER_BINDING                          GL_FRAMEBUFFER_BINDING
```

wm4 replied:
Apple's GL headers being retardedly different from other headers again... But they probably still end up at the same constant value and it's just an annoying warning?

Normally, compilers ignore redundant `#define` statements, which is why our GL headers normally don't clash with system ones. But here Apple decided to define a constant to some other define (because they have the same value or so), which is different from what we do and thus emits a warning.

Can you just avoid including any system GL headers? Otherwise you could add an exception with an ifdef here.

wm4 replied:
Regarding glFlush, you could just store the real function pointer to it in the priv struct.

Akemi replied:
i thought about storing a function pointer in the priv struct too. though i was sure this is was to dirty to do. anyway i pushed the change for this. bloody swift won't let me get function pointers.

wm4 commented on player/client.c:
```diff
@@ -1726,29 +1726,47 @@ int mpv_opengl_cb_uninit_gl(mpv_opengl_cb_context *ctx)
 }
 
-void *mpv_get_sub_api(mpv_handle *ctx, mpv_sub_api sub_api)
+void *mpv_get_sub_api2(mpv_handle *ctx, mpv_sub_api sub_api, bool lock)
```
It doesn't make sense to export this as public API. I think this is not actually exported (would require mpv.def entry), but still the function should not have a `mpv_` prefix to make clear it's private.

Akemi replied:
changed that.

wm4 commented on player/main.c:
```diff
@@ -171,6 +171,10 @@ void mp_destroy(struct MPContext *mpctx)
+#if HAVE_COCOA
+    cocoa_cb_animation_wait();
+#endif
+
```
Can't this be somewhere out of the common code?

Akemi replied:
i would like that too, but we have to block the core somehow till the animation is finished. think there is a good place to put it instead?

wm4 replied:
Why does the core need to be blocked for that?

Akemi replied:
if mpv closes mid animation cocoa can't recover the proper/former state anymore. this leads, in the case of the fs animation, to a always hidden Dock for example.

wm4 replied:
Can't this be hooked to vo_opengl-cb uninit instead?

wm4 replied:
Which mutex/cond?

Akemi replied:
sry, i meant replacing [this](https://github.com/Akemi/mpv/blob/macOS_cocoa-cb2/video/out/cocoa-cb/window.swift#L182) NSCondition with a mutex/cond and moving it to vo_opengl-cb.

Akemi replied:
this is a bit tricky. the events that are triggered at the end of the animation come from the main thread. so waiting in the vo_opengl-cb uninit function deadlocks everything. i could though, start the unlock on a different thread with a delay before the lock, since i know the theoretical duration of the animation. though this a bit hacky and i kinda dislike it.

Akemi replied:
like suggested on irc i could do this differently and remove the waiting from the common code.

wm4 commented on player/main.c:
```diff
@@ -446,6 +450,11 @@ int mp_initialize(struct MPContext *mpctx, char **options)
+        strcmp(opts->vo->video_driver_list->name, "opengl-cb") != 0)
+    {
+        cocoa_cb_dealloc();
+    }
```
I think it's possible for the video_driver_list to have 0 items, but still not be a NULL pointer. Then `name` would be NULL.

Also I'm not so happy that this code exists. I assume it's for switching between the legacy backend?

Akemi replied:
i have to pre-allocate the CAOpeGLLayer because the context is created async by it. i have to make sure cocoa-cb is properly closed if not used, otherwise it would leave an invisible background window, an active context and it would react to events.

maybe there is a better place to call this function, somewhere were we can be sure that opengl-cb is not used?

wm4 replied:
We could add an API function that returns whether opengl-cb is connected to an actual VO. Basically this would return whether `p->ctx->active` is set (see init/uninit functions in vo_opengl_cb.c). The update callback gets called when the VO is destroyed, so you could check this when you receive an update notification.

Akemi replied:
maybe i am misunderstanding something, but the update callback is never called when opengl-cb isn't used? so it would always leave cocoa-cb allocated in that case.

wm4 replied:
Yeah, it has the other way around. We need a "do something if this is not going to be used", which is pretty hard, because we have no idea if it's going to be used.

wm4 commented on video/out/vo_opengl_cb.c:
```diff
@@ -475,18 +508,30 @@ static int control(struct vo *vo, uint32_t request, void *data)
         pthread_mutex_unlock(&p->ctx->lock);
         return VO_TRUE;
+    case VOCTRL_PERFORMANCE_DATA:
+        gl_video_perfdata(p->ctx->renderer, (struct voctrl_performance_data *)data);
```
I don't think this can actually work. Wouldn't you access the GL context from an unrelated thread?

Akemi replied:
you are right that this shouldn't work, looking at the comment in the struct at the top. it does work though besides [this](https://github.com/mpv-player/mpv/issues/4615#issuecomment-326738862). it would be a big pity if we can't that data with cocoa-cb.

wm4 replied:
Well, it's just not sane. It looks like the gl_video call does not actually do any GL calls, which is why this appears to work. But it will still do unsynchronized accesses from 2 threads (the VO thread and the renderer thread). There are two possibilities how to fix this:
1. retrieve the stats after each renderer call, put them into a mutex protected variable, read the copy with VOCTRL (doesn't work, see below)
2. put the performance data into an explicitly shared struct which is protected by its own mutex (and maybe refcounted), which would mean rewriting half the perf reporting code path

Also major WTF: voctrl_performance_data points to allocated strings which the GL code can randomly during rendering! So using this is just a guarantee for undefined behavior (even for normal vp_gpu on other OSes). This mixes undefined behavior with race conditions. This is just insane. I have no idea how I ever (apparently) agreed to such garbage.

So yeah, don't use this, it'll just crash you in obscure situations. If you implement it, it will not only be wrong, but the next person who fixes above bug will have to rewrite your new code too.

Akemi replied:
i had hoped to make stats.lua completely working since it's an embedded feature now. though not much i can do about it.

should a open an issue about this?

wm4 replied:
Yeah. Something about how the VOCTRL is not thread safe. I'm not actually sure how much we even care about this, since apparently the timers are inaccurate anyway.

wm4 commented:
Went over the code. Didn't look at the swift changes. Not to happy about some changes to the common code, but should be easily fixable.

wm4 commented:
Why does the Swift backend need to replace glFlush too? Shouldn't it work like in the GL standard there?

Akemi commented:
because @AirPort still encountered [problems](https://github.com/mpv-player/mpv/issues/4615#issuecomment-326767362) with the early flush.

macdavis commented:
If the window is resized to very small one, during the first part of the fullscreen animation, the aspect ratio is till wrong. It looks like: [https://0x0.st/sPpw.mov](https://0x0.st/sPpw.mov)

Akemi commented:
indeed and fixed. it was a small error. i treated the window aspect ratio as the same as the content/video aspect ratio. though since the titlebar always has a fixed height the window aspect ratio changes with every size change, in the case we have a fixed content/video aspect ratio.

macdavis commented:
Cycling fullscreen/window still has some bugs. Like this, [https://0x0.st/sPDH.m4v](https://0x0.st/sPDH.m4v). Some portion of the image is chopped.

Akemi commented:
i can't reproduce that at all. though it looks like it keeps drawing with the fullscreen aspect ration (with the added black bard from mpv) even so it changed the aspect ratio/size.

would you mind applying this patch and post the terminal output of when that issue happens, pls?

```
diff --git a/video/out/cocoa-cb/video_layer.swift b/video/out/cocoa-cb/video_layer.swift
index 3c16371..f75dbdf 100644
--- a/video/out/cocoa-cb/video_layer.swift
+++ b/video/out/cocoa-cb/video_layer.swift
@@ -105,6 +105,8 @@ class VideoLayer: CAOpenGLLayer {
             surfaceSize!.height *= contentsScale
         }
 
+        print(surfaceSize!)
+
         mpv.drawGLCB(surfaceSize!)
         CGLFlushDrawable(ctx)
 
diff --git a/video/out/cocoa-cb/window.swift b/video/out/cocoa-cb/window.swift
index de625f7..d256afb 100644
--- a/video/out/cocoa-cb/window.swift
+++ b/video/out/cocoa-cb/window.swift
@@ -109,6 +109,8 @@ class VideoWindow: NSWindow, NSWindowDelegate {
             setFrame(frame, display: true)
         }
 
+        Swift.print("changing to fullscreen \(!isInFullscreen)")
+
         if mpv.getBoolProperty("native-fs") {
             super.toggleFullScreen(sender)
         } else {
```

macdavis commented:
I cycled fullscreen/window for several times, the last cycle was the time when the bug appeared.

It seems normally, the terminal prints `(1800.0, 1800.0)
(136.0, 136.0)`, while when the bug appears, it prints `(2880.0, 1800.0)
(136.0, 136.0)`

Full log: [https://0x0.st/sZNl.txt](https://0x0.st/sZNl.txt)

Akemi commented:
@macdavis thx. i think i know what the problem is then. seems like some of the events/notifications for resizing don't always happen at the same time and it locks the rendering size before the fs turns into a window and effectively uses the size o the fullscreen window.

it's a bit annoying that those events are not reliable.

Akemi commented:
@macdavis i pushed a potential fix for this. let me know if this works.

macdavis commented:
Tested around 100 cycles. No problem found. Thanks for your fix.

Akemi commented:
i changed the (un)init routine yet again. it's a lot more robust and sophisticated now. cocoa-cb is now always ready in the background and is not deallocated when not needed. i didn't see much of a resource impact by keeping it in the background. it didn't cause any problems either when the old backend is used. if it starts causing problems i will think about a better solution when it does.

wm4 commented:
Yeah, seems better.

Are the new VOCTRLs really necessary? Can't vo_opengl_cb.c do this?

Akemi commented:
the new VOCTRLs are not necessary if i can get the preinit, uninit and reconfigure event (or the function calls) some other way. tbh i wasn't sure how to do that in a better way. i could move those VOCTRL events to vo_opengl_cb.c but that is probably not what you had in mind. otherwise i would need 3 callbacks for the events or something similar. what would you suggest?

wm4 commented:
Yes, moving them to vo_opengl_cb would be sufficient. Otherwise I'd be afraid other VOs start using them implicitly or something.

Akemi commented:
k, done or would you rather i call control_cb directly instead of using the control function?

wm4 commented:
No, that's probably fine.

A comment on the VOCTRL enum items that they are internal to opengl_cb would probably be nice.

Akemi commented:
indeed i just thought the same.

kevmitch commented on player/client.c:
```diff
@@ -1726,29 +1726,47 @@ int mpv_opengl_cb_uninit_gl(mpv_opengl_cb_context *ctx)
     return MPV_ERROR_NOT_IMPLEMENTED;
 }
+void mp_client_set_control_callback(struct mpv_opengl_cb_context *ctx,
+                                        mpv_opengl_cb_control_fn callback,
```
alignment

kevmitch replied:
alignment

kevmitch commented on video/out/cocoa-cb/window.swift:
```diff
@@ -0,0 +1,456 @@
+            unfsContentFrame = convertToScreen(contentView!.frame)
+            unfScreen = screen
+        }
+        //move window to target screen when going to fullscreen
```
space after //

kevmitch commented on video/out/cocoa-cb/window.swift:
```diff
@@ -0,0 +1,456 @@
+            return newFrame
+        }
+
+        //screen bounds right and left
```
space after //

kevmitch commented on video/out/cocoa-cb/window.swift:
```diff
@@ -0,0 +1,456 @@
+            newFrame.origin.x = targetFrame.origin.x
+        }
+
+        //screen bounds top and bottom
```
space after //

kevmitch commented on video/out/cocoa-cb/window.swift:
```diff
@@ -0,0 +1,456 @@
+        let vf: NSRect = (isAnimating ? targetScreen! : ts).visibleFrame
+        let ncf: NSRect = contentRect(forFrameRect: nf)
+
+        //screen bounds top and bottom
```
space

kevmitch commented on video/out/cocoa-cb/window.swift:
```diff
@@ -0,0 +1,456 @@
+            nf.origin.y = NSMinY(vf) + NSMinY(ncf) - NSMaxY(ncf)
+        }
+
+        //screen bounds right and left
```
space

kevmitch commented on video/out/cocoa_cb_common.swift:
```diff
@@ -0,0 +1,482 @@
+        var geo: vo_win_geometry = vo_win_geometry()
+        vo_calc_window_geometry2(vo, &screenRC, Double(targetScreen.backingScaleFactor), &geo)
+
+        //flip y coordinates
```
space

kevmitch commented on wscript_build.py:
```diff
@@ -139,6 +139,68 @@ def build(ctx):
     ctx(features = "ebml_definitions", target = "ebml_defs.c")
 
+    def swift(task):
+        src = ' '.join(map((lambda x: x.abspath()), task.inputs))
```
`' '.join([x.abspath() for x in task.inputs])`
seems more succinct

kevmitch commented:
in second commit message "because of deadlocking an**d** it needed to be in sync"

kevmitch commented on DOCS/interface-changes.rst:
```diff
@@ -35,6 +35,8 @@ Interface changes
       vf toggle commands and the filter enable/disable flag to customize it.
     - deprecate --af=lavrresample. Use the ``--audio-resample-...`` options to
       customize resampling, or the libavfilter ``--af=aresample`` filter.
+    - deprecate the OpenGL cocoa backend, option choice --gpu-context=cocoa
```
should there be something about this in the man page?

kevmitch commented on DOCS/man/options.rst:
```diff
@@ -5211,6 +5211,9 @@ The following video options are currently all specific to ``--vo=gpu`` and
     flipping GL front and backbuffers immediately (i.e. it doesn't call it
     in display-sync mode).
 
+    On OSX this is always deactivated because it only causes performance
```
Is it really necessary to completely nerf this? Can't it just be off by default?

Akemi replied:
that would either mean to change the default for all platforms or just macOS. former shouldn't be done and latter would mean more platform specific code in the common code, which should be avoided as far as i understand.

i am not quite sure how to handle it then.

kevmitch replied:
ok, fair enough

kevmitch commented on DOCS/interface-changes.rst:
```diff
@@ -70,6 +70,8 @@ Interface changes
       pad must be connected either to another filter, or to a video/audio track
       or video/audio output). If they are disconnected at runtime, the stream
       will probably stall.
+    - deprecate the OpenGL cocoa backend, option choice --gpu-context=cocoa
```
Sorry, I thought I left a comment here yesterday, but it seems to have disappeared. Is it worth noting something about this in the manpage?

kevmitch commented on DOCS/man/options.rst:
```diff
@@ -4862,7 +4862,7 @@ The following video options are currently all specific to ``--vo=gpu`` and
         auto-select (default)
     cocoa
-        Cocoa/OS X
+        Cocoa/OS X (deprecated, use --vo=opengl-cb instead)
```
Sorry, I missed this previously. I was thinking of adding something under the `vo=opengl-cb` entry which currently says "For use with libmpv direct OpenGL embedding; useless in any other contexts."

Akemi replied:
added something there too. if it's fine i will merge that commit.

maru-sama commented:
This fixes #5478 for me.

Akemi commented:
there is one problem i wasn't able to fix. thought i should mention it since it might be a problem that should be considered before merging. when building on 10.12 + xcode9 there will be a linker error `ld: framework not found ColorSync for architecture x86_64`. on 10.12 + xcode8 this doesn't happen (@maru-sama had the same problem). neither does it happen on 10.13 with all the supported xcode (or command line tool) versions. i suspect that's just another bug on Apple's side, probably because of lack of care.

the linker error can be 'fixed' by manually defining the framework path (for what ever reason) like:
```
diff --git a/wscript_build.py b/wscript_build.py
index c9d6299..6aa832b 100644
--- a/wscript_build.py
+++ b/wscript_build.py
@@ -197,6 +197,8 @@ def build(ctx):
         ctx.env.append_value('LINKFLAGS', [
             '-L%s' % _get_swift_static_path(static_swift_paths),
             '-Xlinker', '-force_load_swift_libs', '-lc++',
+            '-F/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks',
+            '-Xlinker', '-framework', '-Xlinker', 'ColorSync',
             '-Xlinker', '-add_ast_path',
             '-Xlinker', '%s' % ctx.path.find_resource("osdep/macOS_swift.swiftmodule")
         ])
```
it even tries to link to some legacy framework `ApplicationServices.framework/Versions/A/Frameworks/ColorSyncLegacy.framework/Versions/A/ColorSyncLegacy`. seems like Apple changed some bits in the frameworks, though it didn't quite work in our case.

anyway after the linker error is fixed i got another error which i couldn't fix.
```
Undefined symbols for architecture x86_64:
  "__dispatch_source_type_data_replace", referenced from:
      __T0So14DispatchSourceC0A0E019makeUserDataReplaceB0So0abdeF0_pSo0A5QueueCSg5queue_tFZ in libswiftDispatch.a(Dispatch.o)
ld: symbol(s) not found for architecture x86_64
```

maru-sama commented:
This version still hangs if you do not use the opengl-cb vo.

Akemi commented:
it doesn't for me. do you have any specific reproduction steps?

maru-sama commented:
If I just use the standard vo cocoa it just hangs during start of the application, no window is displayed. It apparently does not happen if you start it from the CLI.

maru-sama commented:
I think the init sequence is still the culprit here and there is a race somewhere. My new notebook is a lot faster than my old one so this could explain why I am seeing this just now.

Akemi commented:
can you also reproduce it with: `open PATH_TO_MPV/mpv.app --args PATH_TO_FILE`?

maru-sama commented:
No, if I do it like this it starts the file and plays without any delays.

Akemi commented:
does it hang only when double clicking a file or does it also hang when opening mpv directly without a file (eg the drop file here)?

i have a suspect. does this patch help?
diff --git a/video/out/cocoa-cb/video_layer.swift b/video/out/cocoa-cb/video_layer.swift
index c780834..2b890f3 100644
```
--- a/video/out/cocoa-cb/video_layer.swift
+++ b/video/out/cocoa-cb/video_layer.swift
@@ -82,9 +82,9 @@ class VideoLayer: CAOpenGLLayer {
         self.mpv.setGLCBUpdateCallback(self.updateCallback, context: self)
         self.mpv.setGLCBControlCallback(self.cocoaCB.controlCallback, context: self.cocoaCB)
 
-        DispatchQueue.main.async {
+        /*DispatchQueue.main.async {
             self.cocoaCB.updateICCProfile()
-        }
+        }*/
     }
 
     override func canDraw(inCGLContext ctx: CGLContextObj,
```

what does lldb say when you type `bt`?

maru-sama commented:
Commenting this line fixes the issue. of course color correction is disabled then.

Akemi commented:
not completely disable, just the wrong profile on startup (with `icc-profile-auto=yes`). changing profile while running etc still works. though not too important, since it was just for testing if this causes the deadlock.

i will fix this properly a bit later. this is some remain from some old code that should have been changed anyway.

Akemi commented:
that hopefully should fix the deadlock.

maru-sama commented:
@Akemi I can confirm this. It fixes the deadlock with the cocoa vo during startup for me. opengl-cb still works too.

kevmitch commented:
Well I'm not a mac user, so I'm not sure what's best. This PR seems to improve runtime behaviour considerably at the expense of breaking the build in some non-current OS X environments. The biggest problem I see is with people using homebrew for whom this would make mpv uninstallable. Is this trade off worth it? How unusable is mpv without this PR? Can we get by with just telling people to upgrade Apple's broken tools / OS?

Akemi commented:
okay i investigated some more. after a bit of fiddling i got it to build on 10.11 (the oldest OS version i was planning to support). there we have the exact same problem as on 10.12. building with the most recent xcode+sdk doesn't work. though downloading the xcode version that includes the sdk for the OS (eg. xcode7+sdk10.11 on 10.11, same as on 10.12 with xcode8+sdk10.12) it builds and produces a usable binary. this was only a test and the swift code would need to be ported to swift2 (xcode7 only has swift2 and the swift libs are also build with it). i only did a quick fix (ported essential parts to swift 2) for the code so i could build it, so it wasn't completely usable.  i can also confirm that it works with the much smaller command line tools that match the platform. they can be downloaded [here](https://developer.apple.com/download/more/). that's <200mb in comparison to ~5GB for xcode.

my assumption is we will face the same problem on 10.13 whenever xcode10 is released.

it seems like there is a mismatch of linked framework versions and what is included in the sdk and the swift libs. on 10.12 for example when build with xcode9 it can't find ColorSync framework and it actually tries to link to `/System/Library/Frameworks/ApplicationServices.framework/Versions/A/Frameworks/ColorSyncLegacy.framework/Versions/A/ColorSyncLegacy`. this is a framework that changed from 10.12 to 10.13 and also in the respective sdks. the ColorSyncLegacy framework is only available in xcode8 and sdk10.12. i actually also tried to link to those older versions of the frameworks, which only lead to other linking symbol errors. probably obvious because because of different ABIs and what not.

there is probably some magic flag somewhere that fixes it, but the lack of documentation of the swift build tools on macOS is quite a problem in that regard. xcode also wraps those tools in its own tool xcodebuild, as far as i remember. so it is probably a problem to extract the needed information from xcode too.

@kevmitch homebrew could still opt-in to disable it (`--disable-macos-cocoa-cb`) for `--HEAD` builds. i am not quite sure how their bottles work, but that shouldn't be problem when the dev environment is 'properly' set up on their part.

lets say the current cocoa backend is still somewhat usable when you find the right settings for your system to perform 'well', eg combinations of `--no-early-flush` and `--opengl-pbo` most notably. though there are still some annoying issues that could occur depending on system/platform. i personally think it's quite 'unusable', though others might have lower standards. also not every user is well versed, especially on macOS, to find the right settings for themselves. for me this PR is a big jump in usability.

so we can just do one of the following:
1. merge and tell people how to set their dev env up properly to build it
2. fix the underlying problem before merging
3. port everything to obj-c
4. reject this till it is fixed

i will look into 2 a bit longer. maybe i can find something. 3 isn't really an option for me personally since programming in swift was what motivated me quite a bit (sry i am quite tired of obj-c). with 1 i kinda hope that this issue gets a bit of attention and maybe someone smarter comes around and knows how to fix it.

kevmitch commented:
Ok, unless anyone else objects or has any better ideas, I'll let you continue looking at 2 until it drives you sufficiently crazy to give up, at which point just say the word and we'll just merge it and await the onslaught.

kevmitch commented:
Oh, and I guess you should add #5478 to the fixed list?

kevmitch commented:
Just saw your comment on IRC about your simple test program. Is it simple enough to turn into a waf check?

Akemi commented:
@kevmitch added issue #5478 to my initial comment, it was already in the commit message.

test program wasn't too simple but in the end it was just an xcode problem that i could fix after it was driving me nuts all night. anyway i can build that test now which is similar to cocoa-cb in its structure and i will look at all the build stages and linking in the xcode logs. hopefully i can find out why it works there.

i will let you know of my findings.

Akemi commented:
okay, this actually fixes two problems. the problem with this PR and a problem on git master. on git master with 10.12 + SDK10.13 we got linker warnings.
```
ld: warning: text-based stub file /System/Library/Frameworks//QuartzCore.framework/QuartzCore.tbd and library file /System/Library/Frameworks//QuartzCore.framework/QuartzCore are out of sync. Falling back to library file for linking.
ld: warning: text-based stub file /System/Library/Frameworks//AudioToolbox.framework/AudioToolbox.tbd and library file /System/Library/Frameworks//AudioToolbox.framework/AudioToolbox are out of sync. Falling back to library file for linking.
```
we actually need to explicitly set a `-isysroot`, the SDK path, otherwise it tries to link against the system SDK instead of the one it was build for. which obviously leads to linking problems. i can confirm that all linking problems i mentioned before are fixed with this.

i am not sure if there is a better way to solve this problem, since i am not really too knowledgable about our build system. though if it's okay i will merge everything and rebase. then it should be ready for merging.

Akemi commented:
i added a swift version check because the swift3 version shipped with the latest version of xcode on 10.11 doesn't support the "--swift-version" option.

i also reordered the linking fix commit (build: fix linking to dev SDK on macOS), which should probably stay separate since it fixes a not directly related issue, to be before the cocoa-cb commit. after all it is needed for git master and cocoa-cb to link properly.

kevmitch commented on wscript_build.py:
```diff
@@ -139,6 +139,25 @@ def build(ctx):
     ctx(features = "ebml_header", target = "ebml_types.h")
     ctx(features = "ebml_definitions", target = "ebml_defs.c")
 
+    def _run(cmd):
```
Is it possible to move some or all of this to the `waftools/checks/custom.py` so it's done in the configure stage.

kevmitch commented on wscript_build.py:
```diff
@@ -165,13 +165,19 @@ def swift(task):
         header = task.outputs[1].abspath()
         module = task.outputs[2].abspath()
         sflags = ''.join(ctx.env.SWIFT_FLAGS)
+        swift_version = _run('xcrun swift -version').split(' ')[3].split('.')
```
could some this be in `waftools/checks/custom.py` too?

Akemi commented:
okay i moved all build parts that should be in the configure phase to config. i handled the swift compiler on config the same way as `waftools/detections/compiler.py`. i think that is the most appropriate and consistent way?

kevmitch commented:
Looks great. Go ahead and squash. Commit messages should follow [contribute.md](https://github.com/mpv-player/mpv/blob/master/DOCS/contribute.md). Header with prefix and description. Body explaining how / why. Old behaviour before commits is past tense. Otherwise, I think we're just about there.

I can't remember, is there an open issue that can be referenced for the icc-profile deadlock fixes, or was that just a tangential discussion somewhere?

maru-sama commented on DOCS/man/vo.rst:
```diff
@@ -430,7 +430,8 @@ Available video output drivers are:
 ``opengl-cb``
     For use with libmpv direct OpenGL embedding. As a special case, on OS X it
-    is used like a normal VO withih mpv. Otherwise useless in any other contexts.
+    is used like a normal VO withih mpv (cocoa-cb). Otherwise useless in any
```
Small typo there should read "within"

Akemi replied:
fixed

kevmitch commented:
If ever this emoji was justified.

maru-sama commented:
Thank you very much @Akemi for the work you put in that PR and especially for making mpv a first class citizen in the every changing Apple environment again.
See you again when 10.14 hits the streets and Apple decides to change APIs again. :)

rien333 commented:
When I try to upgrade from brew, I get this:

```
$ brew upgrade
...
==> python -c import setuptools... --no-user-cfg install --prefix=/private/tmp/mpv-20180212-94919-k8ftgy/vendor --single-version-externally-managed --record=installed.txt
==> ./bootstrap.py
==> python3 waf configure --prefix=/usr/local/Cellar/mpv/HEAD-c5e4538 --enable-zsh-comp --enable-libmpv-shared --enable-html-build --enable-lua --confdir=/usr/local/etc/mpv --datadir=/usr/local/Cellar/mpv/HEAD-c5e4538/share/mpv --mandir=/
...
File "/private/tmp/mpv-20180212-94919-k8ftgy/waftools/detections/compiler_swift.py", line 17, in __add_swift_flags
    swift_version = __run(ctx.env.SWIFT +  ' -version').split(' ')[3].split('.')
TypeError: can't concat str to bytes
```
Might be an error on their side, but maybe good to know anyway.

maru-sama commented:
Damn, did not test it with python3 in my setup. Can you try if b' -version' fixes this?

Akemi commented:
yeah i am in the process of fixing a few issues.

rien333 commented:
@maru-sama  I would love to, but I'm not sure how to change brew's compilation files. There is a git folder in `/usr/local/Cellar/mpv--git/cache/` or something where I can change the `compiler_swift.py` file, but it seems like brew copies this git cache folder to a different folder in `/tmp` whenever it tries to build from source.

maru-sama commented:
Forget it, this won't work anyway since split is used afterwards which will obviously fail. Akemi is working on it right now.

rien333 commented:
Thanks for fixing the python3 issue for the configuration step! Unfortunately, brew still fails in the linking/compiling step:

```
$ brew upgrade mpv
...
==> ./bootstrap.py
==> python3 waf configure --prefix=/usr/local/Cellar/mpv/HEAD-8762818 --enable-zsh-comp --enable-libmpv-shared --enabl
==> python3 waf install
Last 15 lines from /Users/rw/Library/Logs/Homebrew/mpv/04.python3:
[435/445] Compiling input/event.c
[436/445] Compiling input/ipc.c
[437/445] Compiling misc/thread_pool.c
[438/445] Compiling video/out/opengl/common.c
[439/445] Linking build/mpv
[440/445] Linking build/libmpv.dylib
Undefined symbols for architecture x86_64:
  "_OBJC_CLASS_$__TtC11macOS_swift7CocoaCB", referenced from:
      objc-class-ref in macosx_application.m.26.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Akemi commented:
i believe it's because cocoa-cb is statically linked to the swift libs. for now either build with `--enable-libmpv-static` and not `--enable-libmpv-shared`, or disable cocoa-cb `--disable-macos-cocoa-cb`. i will try to fix that as fast as possible.

rien333 commented:
Okay, thanks for your work. I'll try to edit the homebrew formula to see if I can pass those flags.

Akemi commented:
a quick look, it seems like the problem is that it doesn't try to link to the swift object file when libmpv is linked. maybe my first guess was wrong too.

rien333 commented:
Another question, is it intentional that `$ mpv --vo=help` does not show `opengl-cb`, even though passing it as the vo does work?

Akemi commented:
yeah it's explicitly filtered out [here](https://github.com/mpv-player/mpv/blob/9f595f3a80eeaa0bfbda5702031f054f684f6123/video/out/vo.c#L188). is this a problem and should be changed? it's like this since opengl-cb was only meant to be used with libmpv.

deus0ww commented:
Using vo=opengl-cb, the stats script (shift-i), does not show anything on the second page. This is with MPV HEAD (8762818) with --enable-libmpv-static. Using vo=gpu, the second page works as expected. Is this expected?

Akemi commented:
yes this is a known limitation of opengl-cb. i added the functionality but it was buggy and i had to remove it for now, also see #5322. i will (re)add it with a future PR, i already have a working branch.

AirPort commented:
I've been busy the last couple of days so I couldn't test it in its final version, but now that it's finally been merged I rushed to build it. Unfortunately it seems that commit 235eb60 broke the ffmpeg detection check on my end (macOs 10.13.3). 
Up until that commit it works fine

```
Checking for FFmpeg/Libav present                                    : yes 
Checking for libav* is FFmpeg                                        : yes 
Checking for libav* is Libav                                         : no 
Checking for Libav/FFmpeg library versions                           : yes 
```

since 235eb60 waf gives an error and then it stops

```
Checking for FFmpeg/Libav present                                    : yes 
Checking for libav* is FFmpeg                                        : no 
Checking for libav* is Libav                                         : no 
Checking for Libav/FFmpeg library versions                           : ffmpeg not found 
Unable to find development files for some of the required FFmpeg/Libav libraries. Git master is recommended.
```

(of course it's the latest ffmpeg git version in both cases). Any idea?

Akemi commented:
i will try too look at it tomorrow. seem kinda weird though.

AirPort commented:
If it can be useful, this is the waf [config.log](https://0x0.st/sTHZ.txt) file (and [here's](https://0x0.st/sTHN.txt) the one for the last working commit, abf2efb).

Akemi commented:
@AirPort i see this
> ld: library not found for -ltheoraenc		
> clang: error: linker command failed with exit code 1 (use -v to see invocation)

can you maybe rebuild ffmpeg and/or disable `libtheoraenc`.

AirPort commented:
I'll try and report, but the strange thing about that is that I'm trying to build mpv against the exact same static built ffmpeg's libraries in both cases. 

BTW, what about
```
err: [1m../test.c:3:9: [0m[0;1;31merror: [0m[1m'x' declared as an array with a negative size[0m
{ int x[LIBAVCODEC_VERSION_MICRO >= 100 ? -1 : 1]; return 0; }
[0;1;32m        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[0m[1m../../../../../Applications/mpv-build/build_libs/include/libavcodec/version.h:32:34: [0m[0;1;30mnote: [0mexpanded from macro 'LIBAVCODEC_VERSION_MICRO'[0m
#define LIBAVCODEC_VERSION_MICRO 100
[0;1;32m                                 ^
[0m1 error generated.
```
a few lines beneath?

Akemi commented:
that's the libav check and is supposed to fail, in your case. it also fails in both of your logs.

AirPort commented:
Oh crap, you're right.
Just rebuilt ffmpeg without libtheora though, same error.

Akemi commented:
the same error shouldn't happen if nothing else depends on it or it picks up an older version of ffmpeg. could you maybe try `brew uninstall --force ffmpeg` and rebuilt?

maybe also try removing [line 106-106](https://github.com/mpv-player/mpv/blob/master/waftools/checks/custom.py#L106), the whole. i just tried building on 10.13 too and it worked for me.

AirPort commented:
So, I made it work by removing the the --enable-libsoxr option from ffmpeg build (toghether with --enable-libtheora). I don't know why the presence of those two options suddenly caused the build to fail between those two commit i referenced before, but I honestly think I can live without them. If you still want to investigate more on the matter, I'll be glad to provide any useful information tomorrow.
And to be completely clear on how I build mpv, to maybe better investigate on the issue: while I use brew to install the various dependency, I build mpv and git ffmpeg for it via [mpv-build](https://github.com/mpv-player/mpv-build) (that's why static ffmpeg).

Anyways, good job on the cocoa-cb backend! Seeing all those annoying little issue going away for good is just great.

Akemi commented:
i still can't reproduce the ffmpeg linking issue. my ffmpeg is also build with `libtheora` and `libsoxr`. both are the bottled versions from current homebrew.

deus0ww commented:
To provide another data point, I have no problem installing homebrew ffmpeg with: 
"--with-openssl --with-rtmpdump --with-libass --with-rubberband --with-libsoxr --with-game-music-emu --with-libgsm --with-libvorbis --with-libvpx --with-openjpeg --with-opus --with-speex".

AirPort commented:
Well, I honestly don't care for these two options enough to make you lose your time chasing an elusive issue for basically no gain. 
I'll try again with those option enabled with mpv-build as soon as the build fixes' PR is merged (to make things as clean as possibile), and if it still don't work I'd say to blame static build's shenanigans and just to ignore it unless more people start to report issues about it.

Akemi commented:
if it's a widespread issue i bet we will get a lot more complains from homebrew and more input on the source of the problem. i will probably investigate it at that point some more. sry we only have a workaround for that atm.
